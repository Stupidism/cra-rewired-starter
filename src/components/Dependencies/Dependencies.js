import React from 'react';
import { Link } from 'react-router-dom';

import ExternalLink from 'modules/ExternalLink';
import Loading from 'modules/Loading';
import RandomQuote from 'containers/RandomQuote';
import FetchRandomQuoteButton from 'containers/RandomQuote/FetchRandomQuoteButton';
import LogoSpinSwitch from 'containers/LogoSpinSwitch';

import './Dependencies.less';

const links = {
  someUndefinedRoute: (
    <Link to="/some/undefined/route">some undefined route</Link>
  ),
  reactRouterDom: (
    <ExternalLink to="https://github.com/ReactTraining/react-router">
      react-router-dom
    </ExternalLink>
  ),
  antd: <ExternalLink to="https://ant.design/">antd</ExternalLink>,
  reduxAndReactRedux: (
    <ExternalLink to="https://redux.js.org">redux & react-redux</ExternalLink>
  ),
  recompose: (
    <ExternalLink to="https://github.com/acdlite/recompose">
      recompose
    </ExternalLink>
  ),
  wretch: (
    <ExternalLink to="https://github.com/elbywan/wretch">wretch</ExternalLink>
  ),
  reactLoading: (
    <ExternalLink to="https://codesandbox.io/s/mqx0ql55qp">
      react-loading
    </ExternalLink>
  ),
  reduxThunk: (
    <ExternalLink to="https://github.com/reduxjs/redux-thunk">
      Redux-thunk
    </ExternalLink>
  ),
  lodash: <ExternalLink to="https://lodash.com/">lodash</ExternalLink>,
  reduxCreateReducer: (
    <ExternalLink to="https://www.npmjs.com/package/redux-create-reducer/">
      redux-create-reducer
    </ExternalLink>
  ),
};

const loadingComponent = (
  <div>
    <Loading type="spin" />
    Something you might miss in a blink: above spinning loader is{' '}
    {links.reactLoading}.
  </div>
);

const Dependencies = () => (
  <div className="Dependencies">
    <p>
      Click {links.someUndefinedRoute} to try {links.reactRouterDom}.
    </p>
    <p>
      Look at this beautiful <LogoSpinSwitch /> from {links.antd}. Above buttons
      are from antd, too.
    </p>
    <p>
      You might have noticed, the toggle controls the logo across components
      thanks to {links.reduxAndReactRedux}.
    </p>
    <p>
      Have a look at the toggle's code{' '}
      <code>src/components/Dependencies/LogoSpinSwitch.js</code>,
    </p>
    <p>The code is in HOC-style with the help of {links.recompose}.</p>
    <p>
      About data-fetching, below random famous quote is fetched by{' '}
      {links.wretch}
    </p>
    <div className="QuoteWrapper">
      <RandomQuote loadingComponent={loadingComponent} />
    </div>
    <p>
      {links.reduxThunk} allows async actions like data-fetching and other IO
      operations to work. Try to{' '}
      <FetchRandomQuoteButton>load another quote</FetchRandomQuoteButton>.
    </p>
    <p>
      Also, {links.lodash} and {links.reduxCreateReducer} are used in the
      project for their powerful utility functions.
    </p>
  </div>
);

export default Dependencies;
