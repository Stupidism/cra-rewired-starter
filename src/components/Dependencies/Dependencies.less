.Dependencies {
  .QuoteWrapper {
    width: 720px;
    padding: 15px 25px;
    border: 1px solid #e8e8e8;
    margin: 0 auto 1em;

    .RandomQuote,
    .Loading {
      margin: auto;
    }
  }
}
