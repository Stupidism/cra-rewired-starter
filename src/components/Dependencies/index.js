import Loadable from 'react-loadable';
import delayPromise from 'utils/delayPromise';

import ReactLoadableDependency from '../ReactLoadableDependency';

export default Loadable({
  loader: () =>
    delayPromise(
      import('./Dependencies' /* webpackChunkName: "components.Dependencies" */),
      2000,
    ),
  loading: ReactLoadableDependency,
});
