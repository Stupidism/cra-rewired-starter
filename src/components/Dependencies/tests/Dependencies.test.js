import React from 'react';
import { shallow } from 'enzyme';

import Dependencies from '../Dependencies';

describe('<Dependencies />', () => {
  it('should be consistent with snapshot', () => {
    const snapshot = shallow(<Dependencies />);
    expect(snapshot).toMatchSnapshot();
  });
});
