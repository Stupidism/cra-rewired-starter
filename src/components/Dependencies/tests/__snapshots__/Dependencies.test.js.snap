// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dependencies /> should be consistent with snapshot 1`] = `
<div
  className="Dependencies"
>
  <p>
    Click 
    <Link
      replace={false}
      to="/some/undefined/route"
    >
      some undefined route
    </Link>
     to try 
    <ExternalLink
      to="https://github.com/ReactTraining/react-router"
    >
      react-router-dom
    </ExternalLink>
    .
  </p>
  <p>
    Look at this beautiful 
    <Connect(withHandlers(Component)) />
     from 
    <ExternalLink
      to="https://ant.design/"
    >
      antd
    </ExternalLink>
    . Above buttons are from antd, too.
  </p>
  <p>
    You might have noticed, the toggle controls the logo across components thanks to 
    <ExternalLink
      to="https://redux.js.org"
    >
      redux & react-redux
    </ExternalLink>
    .
  </p>
  <p>
    Have a look at the toggle's code
     
    <code>
      src/components/Dependencies/LogoSpinSwitch.js
    </code>
    ,
  </p>
  <p>
    The code is in HOC-style with the help of 
    <ExternalLink
      to="https://github.com/acdlite/recompose"
    >
      recompose
    </ExternalLink>
    .
  </p>
  <p>
    About data-fetching, below random famous quote is fetched by
     
    <ExternalLink
      to="https://github.com/elbywan/wretch"
    >
      wretch
    </ExternalLink>
  </p>
  <div
    className="QuoteWrapper"
  >
    <Connect(branch(branch(lifecycle(mapProps(RandomQuote)))))
      loadingComponent={
        <div>
          <Loading
            type="spin"
          />
          Something you might miss in a blink: above spinning loader is
           
          <ExternalLink
            to="https://codesandbox.io/s/mqx0ql55qp"
          >
            react-loading
          </ExternalLink>
          .
        </div>
      }
    />
  </div>
  <p>
    <ExternalLink
      to="https://github.com/reduxjs/redux-thunk"
    >
      Redux-thunk
    </ExternalLink>
     allows async actions like data-fetching and other IO operations to work. Try to
     
    <Connect(Component)>
      load another quote
    </Connect(Component)>
    .
  </p>
  <p>
    Also, 
    <ExternalLink
      to="https://lodash.com/"
    >
      lodash
    </ExternalLink>
     and 
    <ExternalLink
      to="https://www.npmjs.com/package/redux-create-reducer/"
    >
      redux-create-reducer
    </ExternalLink>
     are used in the project for their powerful utility functions.
  </p>
</div>
`;
